src/main/java/org/opensearch/flowframework/util/ParseUtils.java

```diff
@@ -228,4 +238,95 @@ public static Map<String, String> getStringToStringMap(Object map, String fieldN
         throw new IllegalArgumentException("[" + fieldName + "] must be a key-value map.");
     }
 
+    /**
+     * Creates a map containing the specified input keys, with values derived from template data or previous node
+     * output.
+     *
+     * @param requiredInputKeys A set of keys that must be present, or will cause an exception to be thrown
+     * @param optionalInputKeys A set of keys that may be present, or will be absent in the returned map
+     * @param currentNodeInputs Input params and content for this node, from workflow parsing
+     * @param outputs WorkflowData content of previous steps
+     * @param previousNodeInputs Input params for this node that come from previous steps
+     * @return A map containing the requiredInputKeys with their corresponding values,
+     *         and optionalInputKeys with their corresponding values if present.
+     *         Throws a {@link FlowFrameworkException} if a required key is not present.
+     */
+    public static Map<String, Object> getInputsFromPreviousSteps(
+        Set<String> requiredInputKeys,
+        Set<String> optionalInputKeys,
+        WorkflowData currentNodeInputs,
+        Map<String, WorkflowData> outputs,
+        Map<String, String> previousNodeInputs
+    ) {
+        // Mutable set to ensure all required keys are used
+        Set<String> requiredKeys = new HashSet<>(requiredInputKeys);
+        // Merge input sets to add all requested keys
+        Set<String> keys = new HashSet<>(requiredInputKeys);
+        keys.addAll(optionalInputKeys);
+        // Initialize return map
+        Map<String, Object> inputs = new HashMap<>();
+        for (String key : keys) {
+            Object value = null;
+            // Priority 1: specifically named prior step inputs
+            // ... parse the previousNodeInputs map and fill in the specified keys
+            Optional<String> previousNodeForKey = previousNodeInputs.entrySet()
+                .stream()
+                .filter(e -> key.equals(e.getValue()))
+                .map(Map.Entry::getKey)
+                .findAny();
+            if (previousNodeForKey.isPresent()) {
+                WorkflowData previousNodeOutput = outputs.get(previousNodeForKey.get());
+                if (previousNodeOutput != null) {
+                    value = previousNodeOutput.getContent().get(key);
+                }
+            }
+            // Priority 2: inputs specified in template
+            // ... fetch from currentNodeInputs (params take precedence)
+            if (value == null) {
```

Comment:
Looking at all the workflow steps, it seems we've always retrieved values from `content` rather than `params`. Can you explain the rationale as to why the parameters take priority here if we've never used params to get values from before?

Reply:
Params are passed by the user on the command line and should generally override other settings. 

Someone could create a template with a default and then change something on the rest path ad hoc. 

Not sure it will ever be used but left it for completeness!